### PR TITLE
Horizontal scaling and status related bug fixes

### PR DESCRIPTION
**Pull request by Mehdi-Bendriss**
### Issue:

This PR fixes some issues that arose by running the `continuous writes` routines:
- during scale up, only recompute the roles once the last unit joins 
- make sure the leader unit's roles can be updated
- for the first unit on the cluster, only wait for the `opensearch service` to start, no need to check the host:port
- better health management
- externalized the health management into separate file/class
- updated charm libs

### Notes:
- Do NOT merge before [this PR](https://github.com/canonical/opensearch-operator/pull/57)

### CODE REVIEW

File: lib/charms/opensearch/v0/opensearch_health.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+        if status == HealthColors.GREEN:
+            # green: cluster healthy
+            self._charm.status.clear(ClusterHealthRed, app=True)
+            self._charm.status.clear(ClusterHealthYellow, app=True)
+            self._charm.status.clear(WaitingForBusyShards, app=True)
+        elif status == HealthColors.RED:
+            # health RED, some primary shards are unassigned
+            self._charm.app.status = BlockedStatus(ClusterHealthRed)
+            return
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_health.py:83]: Why we return None if the status is RED and the actual status in all other cases?

File: lib/charms/opensearch/v0/opensearch_health.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        else:
+            # cluster health is yellow, either a temporary state where shards are moving
+            # around (relocating / initializing) - or a permanent state where some replica
+            # shards are unassigned.
+            self._charm.app.status = (
+                BlockedStatus(ClusterHealthYellow)
+                if status == HealthColors.YELLOW
+                else WaitingStatus(WaitingForBusyShards)
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_health.py:91]: feel free to ignore: If we have a cascade of ifs, for better readability I prefer to have a cascade for all cases (not inline ifs for some cases).

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 3]
```diff
@@ -311,8 +324,8 @@ def _on_opensearch_data_storage_detaching(self, _: StorageDetachingEvent):
             if not self.alt_hosts:
                 raise OpenSearchHAError(NoNodeUpInCluster)
 
-            health = self._apply_cluster_health(wait_for_green_first=True, use_localhost=False)
-            if health == "red":
+            health_color = self.health.apply(use_localhost=False)
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:327]: maybe
```suggestion
            health_color = self.health.apply(wait_for_green_first=True, use_localhost=False)
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: lib/charms/opensearch/v0/opensearch_health.py (hunk 1 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_health.py:83]: Good catch - thank you, `return` should go

File: lib/charms/opensearch/v0/opensearch_health.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        app: bool = True,
+    ) -> str:
+        """Fetch cluster health and set it on the app status."""
+        host = self._charm.unit_ip if use_localhost else None
+        status = self._fetch_status(host, wait_for_green_first)
+
+        if app:
+            self.apply_for_app(status)
+        else:
+            self.apply_for_unit(status)
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_health.py:63]: Good catch, the `app=False` should have been in the `can_service_start`

**Review by WRFitch — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**